pallet/dex/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+	pub type PairOrders<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		(u32, u32),
+		BoundedVec<u64, ConstU32<{ u32::MAX }>>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn user_orders)]
+	pub type UserOrders<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId, //address
+		Blake2_128Concat,
+		u64, //order index,
+		(),
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		SubmitProcessedReceipts {
+			block_number: u64,
+		},
+		OrderCreated {
+			order_index: u64,
+			order: OrderOf<T>,
+		},
+		OrderTaken {
+			account: T::AccountId,
+			order_index: u64,
+			order: OrderOf<T>,
+		},
+		OrderCanceled {
+			order_index: u64,
+		},
+		Deposited {
+			asset_id: u32,
+			amount: u128,
+		},
+
+		Withdrawed {
+			asset_id: u32,
+			amount: u128,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		OrderIndexOverflow,
+		InvalidOrderIndex,
+		InsufficientBalance,
+		NotOwner,
+		AssetIdNotInTokenIndex,
+		AssetIdNotInTokenInfoes,
+		TokenBalanceOverflow,
+		WithdrawBalanceMustKeepOrderSellAmount,
+		UserAssetNotExist,
+		PairOrderNotFound,
+		PairAssetIdMustNotEqual,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight({0})]
+		#[pallet::call_index(0)]
+		pub fn deposit(
+			origin: OriginFor<T>,
+			asset_id: u32,
+			amount: u128,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				TokenIndex::<T>::contains_key(asset_id),
+				Error::<T>::AssetIdNotInTokenIndex
+			);
+
+			<T::Fungibles as Mutate<T::AccountId>>::transfer(
+				asset_id,
+				&who,
+				&Self::account_id(),
+				amount,
+				Preservation::Expendable,
+			)?;
+
+			let mut info = TokenInfo::default();
+			if UserTokenInfoes::<T>::contains_key(who.clone(), asset_id) {
+				info = UserTokenInfoes::<T>::get(who.clone(), asset_id);
+				info.amount = info
+					.amount
+					.checked_add(amount)
+					.ok_or(Error::<T>::TokenBalanceOverflow)?;
+			} else {
+				info.asset_id = asset_id;
+				info.amount = amount;
+			}
+
+			UserTokenInfoes::<T>::insert(who, asset_id, info);
+
+			Self::deposit_event(Event::Deposited { asset_id, amount });
+
+			Ok(().into())
+		}
+
+		#[pallet::weight({1})]
+		#[pallet::call_index(1)]
+		pub fn withdraw(
+			origin: OriginFor<T>,
+			asset_id: u32,
+			amount: u128,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				TokenIndex::<T>::contains_key(asset_id),
+				Error::<T>::AssetIdNotInTokenIndex
+			);
+
+			ensure!(
+				UserTokenInfoes::<T>::contains_key(who.clone(), asset_id),
+				Error::<T>::AssetIdNotInTokenInfoes
+			);
+
+			let mut info = UserTokenInfoes::<T>::get(who.clone(), asset_id);
+			info.amount = info
+				.amount
+				.checked_sub(amount)
+				.ok_or(Error::<T>::TokenBalanceOverflow)?;
+
+			let mut sell_amout = 0;
```

Comment:
Hmm. I think it's complicated logic. 

I think we have to spend balance on orders. This way this behaviour is never happen.

If user put an order, you spend balance. If user cancel order or somebody take order you release this funds. 

User can withdraw only free balance.

@xcthulhu, wdyt?

Reply:
i had add reserved filed
```
pub struct TokenInfo {
	pub amount: u128,
	pub reserved: u128,
}
```